ext/io_gdml.jl

```diff
@@ -135,11 +139,44 @@ function has_volume(e::Ellipsoid{T,<:Any,T}, ::XMLElement, ::XMLElement, id::Int
     end
 end
 
+function has_volume(e::Torus, ::XMLElement, ::XMLElement, id::Integer, pf::AbstractString, v::Bool; parse::Bool = false)
+    if isa(e.r_tube, Tuple)
+        if e.r_tube[1] >= e.r_tube[2]
+            v && @warn "Torus $(pf * string(id)): Outer tube radius must be bigger than the inner tube radius"
+            return false
+        elseif e.r_tube[1] < 0 || e.r_tube[2] < 0
+            v && @warn "Torus $(pf * string(id)): Inner tube radius must be non-negative, Outer tube radius must be positive"
+            return false
+        end
+    elseif e.r_tube <= 0
+        v && @warn "Torus $(pf * string(id)): Tube radius must be a positive number"
+        return false
+    end
+    if e.r_torus <= 0
+        v && @warn "Torus $(pf * string(id)): Torus radius must be a positive number"
+        return true
+    end
+    return true
+end
+
+function has_volume(e::RegularPrism, ::XMLElement, ::XMLElement, id::Integer, pf::AbstractString, v::Bool; parse::Bool = false)
+    if e.r <= 0
+        v && @warn "Prism: $(pf * string(id)): Radius of prism must be a positive number"
+        return false
+    end
+    if e.hZ <= 0
+        v && @warn "Prism: $(pf * string(id)): Height of prism must be a positive number"
+        return false
+    end
+    return true
+end
+
+
 # If composite geometry contains children with zero volume, omit its geometry definition
 # Omit the construction of a Boolean solid if some children have zero volume
-function has_volume(e::AbstractConstructiveGeometry, x_solids::XMLElement, x_define::XMLElement, id::Integer, pf::AbstractString, v::Bool; parse::Bool = true)
-    hva = has_volume(e.a, x_solids, x_define, 2 * id, pf, v, parse = false)
-    hvb = has_volume(e.b, x_solids, x_define, 2 * id + 1, pf, v, parse = false)
+function has_volume(e::AbstractConstructiveGeometry, x_solids::XMLElement, x_define::XMLElement, id::Integer, pf::AbstractString, v::Bool; parse::Bool = false)
```

Comment:
This function does not cover the case for `CSGDifference` if after the difference the volume is zero, e.g. when a bigger volume is subtracted from a smaller volume.
Same thing for a non-overlapping `CSGIntersection`
However, I think that this will never be the case and that we do not have to worry about this.

@bene73 Could you check what happens in these cases and if Geant4 is happy reading in the GDML file or not?

Reply:
Apparently, Geant4 has no problem with volumes that have no points lying inside of them at all (e.g. intersection where $A\cap B=\emptyset$), and only a problem with infinitesimally thin surfaces (e.g. sphere with radius $r=0$).

Thus, intersections could theoretically still cause problems. Two objects can overlap only in an infinitesimally thin surface.

Reply:
Then, I would say we do not have to worry about this!